### ISSUE TITLE
synapser doesn't work in latest Connect 1.7.2

### ISSUE THREAD

**Issue by haozhu233** [Sage-Bionetworks/synapser#247]
## Operating system

centos 7

## Description of the problem

Hi, we are using synapser on our RStudio Connect server in a daily data-downloading script. It works well (on Connect 1.7.0) until we upgraded RStudio Connect to 1.7.2 two days ago. It looks like `PythonEmbededInR` can no longer pick up the correct python installation. I forced `packrat` to rebuild `PythonEmbededInR` by increasing the version number in my forked repo but it didn't work out either. 

I understand this issue might be hard to replicate because RStudio Connect is another commercial product. If you guys also have Connect, you can test it out. I'm contacting rstudio support as well and will update this issue once I hear back.  

### Error log
```
03/25 11:59:46.617
Traceback (most recent call last):
03/25 11:59:46.617
File "<string>", line 1, in <module>
03/25 11:59:46.617
ImportError: No module named pip
03/25 11:59:46.620
Quitting from lines 9-16 (smartphone_update_sql.Rmd)
03/25 11:59:46.620
Error: package or namespace load failed for 'synapser':
03/25 11:59:46.620
.onLoad failed in loadNamespace() for 'PythonEmbedInR', details:
03/25 11:59:46.620
call: value[[3L]](cond)
03/25 11:59:46.620
error: ERROR: Missing system dependencies. Please make sure that your machine has the required dependencies listed in the SystemRequirements field of the DESCRIPTION file: https://github.com/Sage-Bionetworks/PythonEmbedInR/blob/master/DESCRIPTION
```

## Output of `sessionInfo()`
```
R version 3.5.0 (2018-04-23)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1        compiler_3.5.0    prettyunits_1.0.2 bitops_1.0-6     
 [5] base64enc_0.1-3   remotes_2.0.2     tools_3.5.0       testthat_2.0.1   
 [9] digest_0.6.18     packrat_0.5.0     pkgbuild_1.0.3    pkgload_1.0.2    
[13] lattice_0.20-38   jsonlite_1.6      evaluate_0.13     memoise_1.1.0    
[17] rlang_0.3.2       Matrix_1.2-14     cli_1.1.0         rstudioapi_0.10  
[21] curl_3.3          yaml_2.2.0        xfun_0.5          withr_2.1.2      
[25] stringr_1.4.0     knitr_1.22        fs_1.2.7          desc_1.2.0       
[29] askpass_1.1       devtools_2.0.1    grid_3.5.0        rprojroot_1.3-2  
[33] reticulate_1.10   glue_1.3.1        R6_2.4.0          processx_3.3.0   
[37] rmarkdown_1.12    sessioninfo_1.1.1 callr_3.2.0       magrittr_1.5     
[41] usethis_1.4.0     backports_1.1.3   ps_1.3.0          htmltools_0.3.6  
[45] rsconnect_0.8.13  assertthat_0.2.1  stringi_1.4.3     RCurl_1.95-4.12  
[49] openssl_1.3       crayon_1.3.4  
```

**Comment by kimyen** [comment]
Thanks @haozhu233 for reporting the issue.
Please keep us posted if you make progress with RStudio support. Thanks!

**Comment by haozhu233** [comment]
@kimyen yeah, they are looking into this problem. In the mean time I will just use reticulate + your python api.

**Comment by haozhu233** [comment]
Rstudio support got back to me with some initial results. They tested it successful with these two combinations. 
Connect version 1.7.2-7 with R 3.4.4
Connect version 1.7.0-11 with R 3.4.4

The deployment failed, however, with the following combinations due to the embedded python library not building on the test system even with all its dependencies installed:
Connect version 1.7.2-7 with R 3.5.1
Connect version 1.7.2-7 with R 3.5.3
Connect version 1.7.0-11 with R 3.5.3

However, before we upgraded the server, we did get it successful with this package on Connect 1.7.0 with R 3.5.3. This concludes that this problem can somehow be more complicated that I original thought. 🤔

**Comment by kimyen** [comment]
Which repository did you point to to retrieve PythonEmbedInR? We recently switch to an S3 hosted location: http:/ran.synapse.org/ -- The old ran, hosted on Github Pages, sometimes doesn't response well to our current traffic.

**Comment by haozhu233** [comment]
Hi, @kimyen, I think I found out the real cause of this issue! When we deploy the document to Connect, connect is using packrat to do all the version controls for the r packages. We recently installed `reticulate` on our server so it also got installed on Connect together with the package syncing. There might still be a bug somewhere in this pipeline that triggers `reticulate` to load and check for the availability of python even though reticulate was not actually loaded in the document. We see a log:

```
warning: using reticulate but python was not specified; will use python at /usr/bin/python
```
The 2nd half of the log indicates that since we haven't setup a python for `reticulate` to use, it automatically use python at /usr/bin/python. `reticulate` and your `PythonEmbedInR` probably use the same routine in this step so when `PythonEmbedInR` is loaded, it failed because the python source get directed to /usr/bin instead of the /bin folder in the package.

**Comment by kimyen** [comment]
Thank you for helping identifying the causes, @haozhu233 !

We knew about PythonEmbedInR and reticulate not compatible for this particular reason. Why do you need both reticulate and synapser? We do have a [Python package](https://pypi.org/project/synapseclient/) that you can use in reticulate. It's can be installed with `pip3 install synapseclient`. Our synapser just wraps this Python package, so you would expect the syntax to be very similar. Please let me know if this solve the problem for you.

**Comment by haozhu233** [comment]
@kimyen I installed `reticulate` for a different project. In this particular project where we are using synapser to download data, I'm not using reticulate at all. That's why it works on our rstudio server. However, when we publish the rmarkdown document to Connect, packrat detects the existence of the reticulate package and somehow it thinks this document is using reticulate while it's actually not (here is probably where the real bug is). 

Yeah, I was half way through rewriting the script in reticulate when I realize that my synapser code suddenly started to work on Connect. The reason is that in order to get reticulate to work on Connect, I specify `Sys.setenv(RETICULATE_PYTHON = PATH)` in my R profile. This step let us pass the `use_python` check in `reticulate` and avoid the `will use python at /usr/bin/python` step. 

Btw, I just realized that I forgot to post the solution in my previous post. Here it is.
